Bug: `test_counterpoll_watermark.py` fails sporadically due to COUNTERS_DB not being ready

### Is it platform specific

generic

### Importance or Severity

High

### Description of the bug

When running `platform_tests/counterpoll/test_counterpoll_watermark.py` it can fail sporadically with:
```
            if raise_error_when_not_found:
>               raise SonicDbKeyNotFound("No keys for %s found in sonic-db cmd: %s" % (table, cmd))
E               tests.common.helpers.sonic_db.SonicDbKeyNotFound: 'No keys for COUNTERS_PG_*_MAP found in sonic-db cmd:  COUNTERS_DB  keys COUNTERS_PG_*_MAP'
```

This is because the interlock only checks for the presence of any key in `COUNTERS_DB` that matches `COUNTERS_*_*_MAP`
E.G.
```
26/05/2026 19:25:42 sonic_db._run_and_check                  L0040 DEBUG  | SONIC-DB-CLI:  COUNTERS_DB  keys COUNTERS_*_*_MAP
26/05/2026 19:25:42 utilities.wait_until                     L0162 DEBUG  | check_counters_populated is True, exit early with True
...
26/05/2026 19:25:52 sonic_db._run_and_check                  L0040 DEBUG  | SONIC-DB-CLI:  COUNTERS_DB  keys COUNTERS_PG_*_MAP
26/05/2026 19:25:53 sonic_db._run_and_check                  L0044 ERROR  | No command response:  COUNTERS_DB  keys COUNTERS_PG_*_MAP
```

But I polled this counter right after the interlock completed and I can see there are counters there but not ones that would match `COUNTERS_PG_*_MAP`
```
# sonic-db-cli COUNTERS_DB  keys COUNTERS_*_*_MAP
COUNTERS_PORT_NAME_MAP
COUNTERS_SYSTEM_PORT_NAME_MAP
COUNTERS_BUFFER_POOL_NAME_MAP
COUNTERS_LAG_NAME_MAP
COUNTERS_PORT_SERDES_ID_TO_PORT_ID_MAP

# sonic-db-cli COUNTERS_DB  keys COUNTERS_PG_*_MAP

```

After the test failed I can now see the counters present that would've matched `COUNTERS_PG_*_MAP`:
```
# sonic-db-cli COUNTERS_DB  keys COUNTERS_*_*_MAP
COUNTERS_PG_INDEX_MAP
COUNTERS_PG_PORT_MAP
COUNTERS_PORT_SERDES_ID_TO_PORT_ID_MAP
COUNTERS_PG_NAME_MAP
COUNTERS_LAG_NAME_MAP
COUNTERS_PORT_NAME_MAP
COUNTERS_SYSTEM_PORT_NAME_MAP
COUNTERS_DEBUG_NAME_SWITCH_STAT_MAP
COUNTERS_BUFFER_POOL_NAME_MAP

# sonic-db-cli COUNTERS_DB  keys COUNTERS_PG_*_MAP
COUNTERS_PG_INDEX_MAP
COUNTERS_PG_PORT_MAP
COUNTERS_PG_NAME_MAP
```

So if the test waited longer it would've passed.

### Steps to Reproduce

Run `platform_tests/counterpoll/test_counterpoll_watermark.py` many times

### Actual Behavior and Expected Behavior

`platform_tests/counterpoll/test_counterpoll_watermark.py` should wait for the counters it's expecting to show up in `COUNTERS_DB` before continuing.

### Relevant log output

```shell

```

### Output of `show version`

```shell

```

### Attach files (if any)

_No response_

--- a/tests/platform_tests/counterpoll/test_counterpoll_watermark.py
+++ b/tests/platform_tests/counterpoll/test_counterpoll_watermark.py
@@ -65,7 +65,7 @@ def get_keys_on_asics(duthost, db_id, key):
     return {asic.asic_index: SonicDbCli(asic, db_id).get_keys(key) for asic in duthost.asics}
 
 
-def check_counters_populated(duthost, key):
+def check_counters_populated(duthost, key, num_expected_entries):
     """Check that COUNTERS_DB keys matching pattern exist on every ASIC.
 
     The original ``bool(keys.values())`` check was wrong: dict.values()
@@ -80,9 +80,9 @@ def check_counters_populated(duthost, key):
     for asic in duthost.asics:
         try:
             asic_keys = SonicDbCli(asic, "COUNTERS_DB").get_keys(key)
-            if not asic_keys:
+            if len(asic_keys) < num_expected_entries:
                 logging.debug(
-                    "No keys matching '%s' on asic%s yet",
+                    "Not enough keys matching '%s' on asic%s yet",
                     key, asic.asic_index)
                 return False
         except SonicDbKeyNotFound:
@@ -162,11 +162,19 @@ def test_counterpoll_queue_watermark_pg_drop(duthosts, localhost, enum_rand_one_
         timeout = 180 if duthost.is_multi_asic else 120
         pytest_assert(
             wait_until(timeout, 5, delay, check_counters_populated,
-                       duthost, MAPS_LONG_PREFIX.format('*')),
+                       duthost, MAPS_LONG_PREFIX.format('*'), 1),
             "COUNTERS_DB failed to populate after {}s "
             "(counterpoll: {}, method: {}, multi_asic: {})".format(
                 timeout, tested_counterpoll, config_apply_method,
                 duthost.is_multi_asic))
+        # Check each individual counter prefix for the right number of counters in COUNTERS_DB
+        for counters_maps in RELEVANT_MAPS[tested_counterpoll][MAPS]:
+            prefix = counters_maps['prefix']
+            num_expected_counters = len(counters_maps[MAPS])
+            pytest_assert(wait_until(120, 5, 0, check_counters_populated, duthost,
+                                     MAPS_LONG_PREFIX.format(prefix), num_expected_counters),
+                                     "Expected {} entries matching {}".format(
+                                         num_expected_counters, MAPS_LONG_PREFIX.format(prefix)))
     # verify QUEUE or PG maps are generated into COUNTERS_DB after enabling relevant counterpoll
     with allure.step("Verifying MAPS in COUNTERS_DB on {}...".format(duthost.hostname)):
         maps_dict = RELEVANT_MAPS[tested_counterpoll]

--- a/tests/platform_tests/counterpoll/test_counterpoll_watermark.py
+++ b/tests/platform_tests/counterpoll/test_counterpoll_watermark.py
@@ -174,7 +174,7 @@ def test_counterpoll_queue_watermark_pg_drop(duthosts, localhost, enum_rand_one_
             pytest_assert(wait_until(120, 5, 0, check_counters_populated, duthost,
                                      MAPS_LONG_PREFIX.format(prefix), num_expected_counters),
                                      "Expected {} entries matching {}".format(
-                                         num_expected_counters, MAPS_LONG_PREFIX.format(prefix)))
+                                     num_expected_counters, MAPS_LONG_PREFIX.format(prefix)))
     # verify QUEUE or PG maps are generated into COUNTERS_DB after enabling relevant counterpoll
     with allure.step("Verifying MAPS in COUNTERS_DB on {}...".format(duthost.hostname)):
         maps_dict = RELEVANT_MAPS[tested_counterpoll]

--- a/tests/platform_tests/counterpoll/test_counterpoll_watermark.py
+++ b/tests/platform_tests/counterpoll/test_counterpoll_watermark.py
@@ -173,8 +173,8 @@ def test_counterpoll_queue_watermark_pg_drop(duthosts, localhost, enum_rand_one_
             num_expected_counters = len(counters_maps[MAPS])
             pytest_assert(wait_until(120, 5, 0, check_counters_populated, duthost,
                                      MAPS_LONG_PREFIX.format(prefix), num_expected_counters),
-                                     "Expected {} entries matching {}".format(
-                                     num_expected_counters, MAPS_LONG_PREFIX.format(prefix)))
+                          "Expected {} entries matching {}".format(
+                              num_expected_counters, MAPS_LONG_PREFIX.format(prefix)))
     # verify QUEUE or PG maps are generated into COUNTERS_DB after enabling relevant counterpoll
     with allure.step("Verifying MAPS in COUNTERS_DB on {}...".format(duthost.hostname)):
         maps_dict = RELEVANT_MAPS[tested_counterpoll]

